Check all documents specified in Product_Document tags for correct mi…

qchaupds opened the pull request:
…me types.

1. Add new test resources :src/test/resources/github367
2. Add needed dependency for group com.sun.activation, which is the implementation jar :pom.xml
3. Add default mime type resource file :src/main/resources/validate_default_mime_types.txt
4. Add new class to validate JPEG and PNG image file content :src/main/java/gov/nasa/pds/tools/util/ImageUtil.java
5. Add new class to check if the mime type of a given file matches its file extension :src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java
6. Add new problem types :src/main/java/gov/nasa/pds/tools/validate/ProblemType.java
7. Add check for all file_name tags in Product_Document tags :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java
8. Add new regression test :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/367 As a user, I want to validate all files referenced by a Product_Document

Previously, validate module only check for the correct file format of PDF.  Now, the code checks for all file names
for matching mime types.  Code also check for correctness of JPEG and PNG file content.
Note that this does not check for correct file content for the other un-mentioned types.

Tested in DEV:

Case 1:
Running validate against a bundle with document label with mime types incorrect.

```
validate -R pds4.bundle -r report_github367_bundle_valid.json  -s json -t src/test/resources/github367/bundle_kaguya_derived.xml >& t3
```

There should be 21 warning messages, with 12 messages related to the wrong mime types.

```
{pds-dev3.jpl.nasa.gov}/home/qchau/sandbox/validate 213 % grep "not correct" report_github367_bundle_valid.json | cat -n
     1            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_2.txt with document standard id 'Microsoft Word' is not correct"
     2            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.html with document standard id 'MPEG-4' is not correct"
     3            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.ps with document standard id 'HTML' is not correct"
     4            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.eps with document standard id 'TIFF' is not correct"
     5            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.gif with document standard id 'TIFF' is not correct"
     6            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.doc with document standard id 'JPEG' is not correct"
     7            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.tex with document standard id 'GIF' is not correct"
     8            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.mp4 with document standard id 'PDF' is not correct"
     9            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.xls with document standard id 'Encapsulated Postscript' is not correct"
    10            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.doc with document standard id 'HTML' is not correct"
    11            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.rtf with document standard id 'LaTEX' is not correct"
    12            "message": "file:/home/qchau/sandbox/validate/src/test/resources/github367/document/pccref_empty.tif with document standard id 'Microsoft Excel' is not correct"
```

Commit:
Check all documents specified in Product_Document tags for correct mime types.

1. Add new test resources :src/test/resources/github367
2. Add needed dependency for group com.sun.activation, which is the implementation jar :pom.xml
3. Add default mime type resource file :src/main/resources/validate_default_mime_types.txt
4. Add new class to validate JPEG and PNG image file content :src/main/java/gov/nasa/pds/tools/util/ImageUtil.java
5. Add new class to check if the mime type of a given file matches its file extension :src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java
6. Add new problem types :src/main/java/gov/nasa/pds/tools/validate/ProblemType.java
7. Add check for all file_name tags in Product_Document tags :src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java
8. Add new regression test :src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/367 As a user, I want to validate all files referenced by a Product_Document

Commit:
Using alternative method to get the parent

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:
```diff
@@ -0,0 +1,204 @@
+    LOG.debug("isPNG:pngBase {}",pngBase);
+
+    // Get the location of the input file.
+    URI uri = null;
```
@sonatype-lift ignore

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:
```diff
@@ -0,0 +1,204 @@
+    try {
+        // Read the first 8 bytes.
+        int firstInt = 0;
+        int secondInt = 0;
```
@sonatype-lift ignore

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:
```diff
@@ -0,0 +1,204 @@
+    try {
+        // Read the first 2 bytes and last 2 bytes.  If they are 0xffd8 and 0xffd9 then the file is a JPEG file.
+        int firstShort = 0; 
+        int lastShort = 0; 
```
@sonatype-lift ignore

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:
```diff
@@ -0,0 +1,204 @@
+
+    try {
+        // Read the first 2 bytes and last 2 bytes.  If they are 0xffd8 and 0xffd9 then the file is a JPEG file.
+        int firstShort = 0; 
```
@sonatype-lift ignore

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/ImageUtil.java:
```diff
@@ -0,0 +1,204 @@
+
+    try {
+        // Read the first 8 bytes.
+        int firstInt = 0;
```
@sonatype-lift ignore

Commit:
Fix sonatype-lift complaints: unused variables

jordanpadams commented on src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java:
```diff
@@ -350,6 +426,52 @@ private boolean validate(DocumentInfo xml) {
+                  passFlag = false;
+                }
+
+                passFlag = this.checkGenericDocument(target, urlRef, fileObject, htmlName, parent, directory, documentStandardId, "HTML", problems, ProblemType.NON_HTML_FILE);
```
@qchaupds it seems like this should be a loop through a hashmap, but not worth holding up the PR right now.

jordanpadams reviewed (changes requested):
@qchaupds looks like these tests are failing. can you please fix ASAP and I will merge?

Commit:
Updates to provide an alternative to MimetypesFileTypeMap to find a mime type

1. New class to get a mime type given a file name (an alternative to MimetypesFileTypeMap) : src/main/java/gov/nasa/pds/tools/util/MimeTable.java
2. Add logic to switch between the 2 methods of getting mime type : src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java
3. Fix bug by correcting parameter to checkGenericDocument for MSEXCEL document type : src/main/java/gov/nasa/pds/tools/validate/rule/pds4/FileReferenceValidationRule.java

Refs:

https://github.com/nasa-pds/validate/issues/367 As a user, I want to validate all files referenced by a Product_Document

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/MimeTable.java:
```diff
@@ -0,0 +1,151 @@
+             
+          }
+      }
+      LOG.debug("loadMimeTable:defaultMimeTypesFileName,mimetypesFileTypeMap.size {},{}",defaultMimeTypesFileName,mimetypesFileTypeMap.size());
```
@qchaupds let's definitely check this leak if we can.

Commit:
Fixes to sonatype complaints and add important missing resource file to pom.xml

1. Few fixes to sonatype complaints : src/main/java/gov/nasa/pds/tools/util/MimeTable.java
2. Add important missing validate_default_mime_types.txt resource so it can be found to includes tag pom.xml
3. Correct location of resource file validate_default_mime_types.txt : src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java

Refs:

https://github.com/nasa-pds/validate/issues/367 As a user, I want to validate all files referenced by a Product_Document

qchaupds commented on src/main/java/gov/nasa/pds/tools/util/MimeTable.java:
```diff
@@ -0,0 +1,178 @@
+          LOG.debug("loadMimeTable:FilenameUtils.getName(defaultMimeTypesFileName) {}",FilenameUtils.getName(defaultMimeTypesFileName));
+          LOG.debug("loadMimeTable:FilenameUtils.getPath(defaultMimeTypesFileName) {}",FilenameUtils.getPath(defaultMimeTypesFileName));
+          LOG.debug("loadMimeTable:parent {}",parent);
+          inFile = new Scanner(new File(parent + File.separator + FilenameUtils.getName(defaultMimeTypesFileName)));
```
@sonatype-lift ignore  The name has already been sanitized by fetching the parent plus the filename separately.

Commit:
Swap in a better alternative to resolve a system resource

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/DocumentsChecker.java:
```diff
@@ -0,0 +1,133 @@
+  private static final String POSTSCRIPT_ENDS_WITH_PATTERN_1 = ".ps";
+  private static final String POSTSCRIPT_ENDS_WITH_PATTERN_2 = ".PS";
+  private static final String ENCAPSULATED_POSTSCRIPT_ENDS_WITH_PATTERN_1 = ".eps";
+  private static final String ENCAPSULATED_POSTSCRIPT_ENDS_WITH_PATTERN_2 = ".EPS";
```
@sonatype-lift ignore

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/MimeTable.java:
```diff
@@ -0,0 +1,151 @@
+      LOG.debug("loadMimeTable:defaultMimeTypesFileName {}",defaultMimeTypesFileName);
+      Scanner inFile = null;
+      try {
+          inFile = new Scanner(new File(defaultMimeTypesFileName));
```
@sonatype-lift help

jordanpadams committed:
Removing unused variable

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/MimeTable.java:
```diff
@@ -0,0 +1,151 @@
+              // 7-Bit_ASCII_Text txt text TXT TEXT
+              tokens = nextLine.split("\\s+");
+
+              LOG.debug("loadMimeTable:nextLine,tokens.length [{}],{}",nextLine,tokens.length);
```
@sonatype-lift ignore

jordanpadams commented on src/main/java/gov/nasa/pds/tools/util/MimeTable.java:
```diff
@@ -0,0 +1,151 @@
+
+              tokenIndex = 1;
+              while (tokenIndex < tokens.length) {
+                  LOG.debug("loadMimeTable:APPENDING_KEY {},{},[{}],[{}]",tokenIndex,tokens.length,tokens[tokenIndex],tokens[0]);
```
@sonatype-lift ignore
